### ISSUE TITLE
Websocket how to get size of frame with ReceiveAsync?

### ISSUE THREAD

**Issue by ArjunVachhani** [dotnet/runtime#35174]
WebSocket.ReceiveAsync returns the chunk/part of the web socket frame. ReceiveAsync should also expose the payload size of the frame. Current implementation returns Count, MessageType, and EndOfMessage. while allocating memory, it would really helpful knowing the size of the frame.

**Comment by davidsh** [comment]
>It would really helpful knowing the size of the frame.

What scenario do you have where this information would be useful?

**Comment by ArjunVachhani** [comment]
clients can send data variable in length. allocating less or more memory leads to pressure on garbage collector.  below is sample code to read a whole frame from web-socket.


```csharp
var framePayload = new Memory<byte>(new byte[4 * 1000]);// <-- how to allocate efficiently
int framePayloadSize = 0;
do
{
	var buffer = new Memory<byte>(new byte[1000]);
	result = await socket.ReceiveAsync(buffer, cancellationTokenSource.Token);
	var data = buffer.Slice(0, result.Count);
	data.CopyTo(framePayload.Slice(framePayloadSize, result.Count));
	framePayloadSize += result.Count;
} while (!result.EndOfMessage);
```

If there is a property which tells size of frame then while reading first chunk of a frame we can initialize framePayload with correct size.

**Comment by davidfowl** [comment]
This would actually be a very nice API to add. It would require you to incur another API call though (to basically read the frame header).  

- If we exposed another API to get a header information, we'd need to figure out how it would interact with the existing API and what misuse looks like.
- Would this API be used before every single read, or once per logical message (end of message true)? 

@ArjunVachhani Can you say how you'd want to use this API? What would your ideal code look like?

cc @BrennanConroy

**Comment by ArjunVachhani** [comment]
Right now `MessageType` frame level property is exposed via `WebSocketReceiveResult` class, so I think you could either expose frame payload size in `WebSocketReceiveResult` or expose a method/property on `WebSocket` which will return frame payload size and `MessageType` property.

**Comment by davidfowl** [comment]
> Right now MessageType frame level property is exposed via WebSocketReceiveResult class, so I think you could either expose frame payload size in WebSocketReceiveResult or expose a method/property on WebSocket which will return frame payload size and MessageType property.

Can you write out some pseudo code for what this would look like? Also what do you think about the questions above? What would happen if you call it at the wrong time (ReceiveHeader twice in a row for example, does it return the same value until the message is fully consumed?)

**Comment by ArjunVachhani** [comment]
I can think of 2 ways to read from `WebSocket`. Option 1 is closer to current implementation. Option 2 looks easier for developer and is probably more efficient as it would not require copying data around.

**Option 1**
```csharp 
while (socket.State == WebSocketState.Open)
{
	int chunksRecieved = 0;
	int framePayloadSize = 0;
	Memory<byte> framePayload = null;
	var buffer = new Memory<byte>(new byte[1000]);
	do
	{
		result = await socket.ReceiveAsync(buffer, cancellationTokenSource.Token);

		if (chunksRecieved == 0)
			framePayload = new Memory<byte>(new byte[result.PayloadSize]); //initialize with proper size

		var data = buffer.Slice(0, result.Count);
		data.CopyTo(framePayload.Slice(framePayloadSize, result.Count));
		framePayloadSize += result.Count;
		chunksRecieved++;
	} while (!result.EndOfMessage);

	//process frame here
}
```
**Option 2**
```csharp
while (socket.State == WebSocketState.Open)
{
	int framePayloadReceived = 0;
	var headerResult = await socket.ReceiveHeaderAsync(cancellationTokenSource.Token); //this will wait until frame is received. headerResult will hold MessageType & PayloadSize
	Memory<byte> framePayload = new Memory<byte>(new byte[headerResult.PayloadSize]); //initialize with proper size.
	do
	{
		var buffer = framePayload.Slice(framePayloadReceived, framePayload.Length - framePayloadReceived);
		result = await socket.ReceiveAsync(buffer, cancellationTokenSource.Token);
		framePayloadReceived += result.Count;
	} while (!result.EndOfMessage);
	//process frame here
}
```

ReceiveHeaderAsync should probably return header of current frame. If there no pending frame to be read then it should wait. calling it multiple times should return same result for same frame.

Please let me know if i missed/misunderstood anything, 

cheers :)

**Comment by davidfowl** [comment]
@stephentoub Thoughts? I haven't thought through what happens when you get continuation frames but that should work I guess.

**Comment by stephentoub** [comment]
It seems reasonable.  The biggest roadblock is probably that `WebSocket` is abstract, and there's no good way to implement this on top of any of the existing APIs, so a new `ReceiveHeaderAsync` method would need to be `virtual` and probably throw a `NotSupportedException`.  It could then be implemented by the various implementations, e.g. `ManagedWebSocket` as returned from `WebSocket.CreateFromStream`, but there might also need to be a capability API `SupportsReceiveHeaderAsync` (not only because of the abstract thing, but also because I expect we might have trouble implementing this API on top of other websockets OS APIs when necessary). We'd also need to work out the semantics for how ReceiveHeaderAsync behaves when we've split a frame due to a previous ReceiveAsync call not providing a large enough buffer to store the whole thing or us not having received the whole payload yet.

**Comment by karelz** [comment]
Triage: We would be open to take this if we have proof of concept -- @ArjunVachhani is it something you would be interested in helping us with?

**Comment by zlatanov** [comment]
Now that WebSockets support compression, I want to point out that this API is incompatible (or at least makes no sense) when the message is compressed. The reason for this is that the payload length in this case represents the compressed data, but the user actually wants to know the uncompressed (final) size.

@davidfowl what do you think?

**Comment by halter73** [comment]
I agree compression is a snag. What are the options? Dry run inflate to determine the output buffer size? Seems wasteful. Don't support this for compressed WebSockets? I'm not an deflate expert so maybe there's an easy option I'm not aware of.

**Comment by zlatanov** [comment]
@halter73 Dry run isn't possible, because it would have to receive everything and decompress it using private buffers, which would defeat the purpose of this API - knowing how much buffer size we need to receive the message in the first place.

Conditionally supporting this seems wrong to me, because the name of the method `ReceiveHeaderAsync` doesn't give the impression it would not work if the next message is compressed.

What seems possible however is that the `length` information returned from this method to be nullable. Something like:

```c#
readonly struct WebSocketMessageHeader
{
    public WebSocketMessageType MessageType { get; }
    public bool IsFragment { get; }
    public bool IsCompressed { get; }
    public long? PayloadLength { get; } // null when IsCompressed == true
}
```

**Comment by zlatanov** [comment]
> does the WebSocket decompress the message before it returns?

Yes. The payload is automatically decompressed before returned to the user.

**Comment by MeTitus** [comment]
> This would actually be a very nice API to add. It would require you to incur another API call though (to basically read the frame header).
> 
> * If we exposed another API to get a header information, we'd need to figure out how it would interact with the existing API and what misuse looks like.
> * Would this API be used before every single read, or once per logical message (end of message true)?
> 
> @ArjunVachhani Can you say how you'd want to use this API? What would your ideal code look like?
> 
> cc @BrennanConroy

Well this code snipped is obviously wrong but then again so is the inner implementation. It leaves it open for messages to received different WebSocketMessageType which should not happen.

**Comment by MeTitus** [comment]
> @halter73 Dry run isn't possible, because it would have to receive everything and decompress it using private buffers, which would defeat the purpose of this API - knowing how much buffer size we need to receive the message in the first place.
> 
> Conditionally supporting this seems wrong to me, because the name of the method `ReceiveHeaderAsync` doesn't give the impression it would not work if the next message is compressed.
> 
> What seems possible however is that the `length` information returned from this method to be nullable. Something like:
> 
> ```cs
> readonly struct WebSocketMessageHeader
> {
>     public WebSocketMessageType MessageType { get; }
>     public bool IsFragment { get; }
>     public bool IsCompressed { get; }
>     public long? PayloadLength { get; } // null when IsCompressed == true
> }
> ```

Does it seem ok to you then, that the same data split in two different messages can use different WebSocketMessageType? That should not be left for clients to manage. Obviously the right right is to have a leading packages and payload ones which belong to the same initial leading one.

**Comment by cittaz** [comment]
Hello, I just stumbled upon this stale suggestion.

I'd like to revive this issue and ask whether **Option 1** is still considered viable: exposing the incoming payload size directly on the receive result, for example as `PayloadSize` / `PayloadLength`.

I think this could be useful for more than just convenience. A concrete case is when we issue a zero-byte receive first so idle connections do not allocate a buffer while waiting for data, as in the ASP.NET Core SignalR WebSocket server transport:

```csharp
var result = await socket.ReceiveAsync(Memory<byte>.Empty, token);

if (result.MessageType == WebSocketMessageType.Close)
{
    _gracefulClose = true;
    return;
}

var memory = _application.Output.GetMemory();

var receiveResult = await socket.ReceiveAsync(memory, token);
```

If the first receive result exposed the payload size, the transport could pass that as a `PipeWriter.GetMemory(sizeHint)` hint:

```csharp
var result = await socket.ReceiveAsync(Memory<byte>.Empty, token);

if (result.MessageType == WebSocketMessageType.Close)
{
    _gracefulClose = true;
    return;
}

var memory = _application.Output.GetMemory(result.PayloadSize); // Maybe clamped somehow

var receiveResult = await socket.ReceiveAsync(memory, token);
```

The important part is not necessarily allocating exactly the full payload size. Callers could, and probably should, clamp the value. But even a bounded size hint could let `PipeWriter` choose a better segment size than the current no-hint path.

This seems especially relevant for code that wants to keep the current allocation-friendly pattern:

1. wait for incoming data with a 0-byte receive;
2. avoid allocating while idle;
3. once a frame/message is known to be available, ask the destination writer for a suitably sized buffer;
4. receive directly into that buffer.

A separate API, as in Option 2, may still work, but it feels less natural for this case because the information is already discovered as part of receive processing. Having it on the result would compose better with existing `ReceiveAsync` loops.

A few open problems that might help move the API discussion forward if discussed:

- `ValueWebSocketReceiveResult` would likely increase in size because we would probably need to add an `int` field.
- When compression is enabled, the API might not work as expected, so a decision should be made on how to set the value in that case.

Note: the original comment with code snippets for proposals of Option 1 and Option 2 is likely incomplete because that code would not work when continuation frames are received and the whole message does not arrive in a single frame. Still, I consider the two options valid and usable in some contexts.

What's your take? I'm open to work on this if you think it make sense!

cc 
@karelz @MihaZupan
@davidfowl (I saw you were happy about the original proposal + the aspnet case may be relevant for you)
